src/main/python/deep_qa/layers/backend/batch_dot.py

```diff
@@ -0,0 +1,120 @@
+import keras.backend as K
+from keras.layers import Layer
+from overrides import overrides
+
+
+class BatchDot(Layer):
+    """
+    This ``Layer`` calls ``K.batch_dot()`` on two inputs ``tensor_a`` and ``tensor_b``.
+    This function will work for tensors of arbitrary size as long as
+    ``abs(K.ndim(tensor_a) - K.ndim(tensor_b)) < 1``, due to limitations in ``K.batch_dot()``.
+
+    We always assume the dimension to perform the dot is the last one, and that
+    the masks have one fewer dimension that the tensors. Note that this layer
+    does not return zeroes in places that are masked, but does pass a correct
+    mask forward. If this then gets fed into ``masked_softmax``, for instance,
+    your tensor will be correctly normalized. We always assume the dimension to
+    perform the dot is the last one, and that the masks have one fewer
+    dimension than the tensors.
+
```

Comment:
Some more documentation here about what happens with various example shapes would be good.